ansible/roles/vm_set/tasks/add_topo.yml

```
@@ -19,17 +19,34 @@
     container_type: "API-SERVER"
   when: ptf_imagename is defined and ptf_imagename == "docker-keysight-api-server"
 
+- name: Try to pull ptf docker image
+  docker_image:
+    name: "{{ docker_registry_host }}/{{ ptf_imagename }}:{{ ptf_imagetag }}"
+    source: pull
+    build:
+      pull: yes
+  become: yes
+  register: docker_pull_result
+  ignore_errors: yes
+
+- name: Try to login into docker registry if pull failed
+  docker_login:
+    registry_url: "{{ docker_registry_host }}"
+    username: "{{ docker_registry_username }}"
+    password: "{{ docker_registry_password }}"
+  become: yes
+  when: docker_pull_result.failed == true
```

Is it feasible to do something like this before the pull:
```suggestion
- name: Try to login into docker registry if pull failed
  docker_login:
    registry_url: "{{ docker_registry_host }}"
    username: "{{ docker_registry_username }}"
    password: "{{ docker_registry_password }}"
  become: yes
  when: docker_registry username is defined and docker_registry_password is defined
```

---

Good suggestion! Thanks, I'll update.